rpcserver.go

```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
 	return nil, nil
 }
 
+func handleAddMiningAddr(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
+	c := cmd.(*btcjson.AddMiningAddrCmd)
+
+	// Decode the addresses encoded as a string into an in-memory object.
+	addr, err := btcutil.DecodeAddress(c.Address, activeNetParams.Params)
+	if err != nil {
+		return nil, err
+	}
+
+	// Scan the current list of mining addresses to ensure that we don't
+	// add any duplicates.
+	cfg.miningAddrMtx.RLock()
+	for _, miningAddr := range cfg.miningAddrs {
+		if miningAddr.String() == addr.String() {
+			cfg.miningAddrMtx.RUnlock()
+			return nil, &btcjson.RPCError{
+				Code:    btcjson.ErrRPCInternal.Code,
+				Message: "Duplicate mining address detected",
+			}
+		}
+	}
+	cfg.miningAddrMtx.RUnlock()
+
+	// If adding this new address will not introduce any duplicate mining
+	// addresses, then append it to the list of available mining addresses.
+	cfg.miningAddrMtx.Lock()
+	cfg.miningAddrs = append(cfg.miningAddrs, addr)
+	cfg.miningAddrMtx.Unlock()
+
+	return nil, nil
+}
+
+func handleDelMiningAddr(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
+	c := cmd.(*btcjson.DelMiningAddrCmd)
+
+	// Decode the addresses encoded as a string into an in-memory object.
+	addr, err := btcutil.DecodeAddress(c.Address, activeNetParams.Params)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.miningAddrMtx.Lock()
+	for i := 0; i < len(cfg.miningAddrs); i++ {
+		if cfg.miningAddrs[i].String() == addr.String() {
+			// Delete the element in-place from the slice of mining
+			// addresses by sliding elements to the right of the
+			// index over one.
+			copy(cfg.miningAddrs[i:], cfg.miningAddrs[i+1:])
+
+			// Then we set the final element to a nil value in
+			// order to avoid of GC leak, then re-slice the slice
+			// to reduce its length by one (cutting off the newly
+			// added nil element).
+			cfg.miningAddrs[len(cfg.miningAddrs)-1] = nil
+			cfg.miningAddrs = cfg.miningAddrs[:len(cfg.miningAddrs)-1]
+
+			cfg.miningAddrMtx.Unlock()
+			return nil, nil
+		}
+	}
+	cfg.miningAddrMtx.Unlock()
+
+	// If we've reached this point, then the targeted mining address wasn't
+	// found so we return an error indicating so.
+	return nil, &btcjson.RPCError{
+		Code:    btcjson.ErrRPCInternal.Code,
```

Comment:
This isn't an internal error.  It was an invalid address provided by the caller.  Thus, it should be `btcjson.ErrRPCInvalidAddressOrKey`.